core/src/main/scala/org/http4s/StaticFile.scala

```diff
@@ -87,11 +88,18 @@ object StaticFile {
             else `Transfer-Encoding`(TransferCoding.chunked)
           val headers = Headers(lenHeader :: lastModHeader ::: contentType)
 
-          Some(
-            Response(
-              headers = headers,
-              body = readInputStream[F](F.delay(url.openStream), DefaultBufferSize, blocker)
-            ))
+          Try(url.openStream()).fold(
+            fa = {
+              case _: FileNotFoundException => None
```

Comment:
I guess this may be unsafe – if it happens that `openStream` will throw some other exception, then we get `MatchError` here.

Reply:
@satorg I don't thing there is a way to have `utl.openStream` in `F.delay`. For example if I put `url.openStream` in `F.delay`

```
val r: F[Option[Stream[F,Byte]]] = F.delay(url.openStream)
            .redeem(recover = {
              case _: FileNotFoundException => None
            }, f = {inputStream => Some(readInputStream[F](F.pure(inputStream), DefaultBufferSize, blocker))})
```
Notice that the type of `r` is wrapped in a `F`. But we want a result of the type `Option[Response[_][F]]`.

Reply:
@ashwinbhaskar it should be almost always possible :)
This implementation works for me (as well as against the test case in your PR):
```scala
  def fromURL[F[_]](url: URL, blocker: Blocker, req: Option[Request[F]] = None)(implicit
      F: Sync[F],
      cs: ContextShift[F]): OptionT[F, Response[F]] = {
    val fileUrl = url.getFile()
    val file = new File(fileUrl)
    OptionT.apply(F.suspend {
      if (file.isDirectory()) // TODO: this call should be suspended in the blocker
        F.pure(None)
      else {
        val urlConn = url.openConnection
        val lastmod = HttpDate.fromEpochSecond(urlConn.getLastModified / 1000).toOption
        val ifModifiedSince = req.flatMap(_.headers.get(`If-Modified-Since`))
        val expired = (ifModifiedSince, lastmod).mapN(_.date < _).getOrElse(true)

        if (expired) {
          val lastModHeader: List[Header] = lastmod.map(`Last-Modified`(_)).toList
          val contentType = nameToContentType(url.getPath).toList
          val len = urlConn.getContentLengthLong
          val lenHeader =
            if (len >= 0) `Content-Length`.unsafeFromLong(len)
            else `Transfer-Encoding`(TransferCoding.chunked)
          val headers = Headers(lenHeader :: lastModHeader ::: contentType)

          blocker
            .delay(url.openStream)
            .redeemWith[Option[Response[F]]](
              {
                case _: FileNotFoundException => F.pure(None)
                case other => F.raiseError(other)
              },
              stream =>
                F.pure(
                  Some(
                    Response[F](
                      headers = headers,
                      body = readInputStream[F](stream.pure[F], DefaultBufferSize, blocker))))
            )
        } else
          blocker.delay(urlConn.getInputStream.close()) >>
            F.pure(Some(Response(NotModified)))
      }
    })
  }
```
Feel free to borrow any part of this snippet if you'd like to.

Reply:
ah..I see the `suspend` did the trick:)